src/orders/views/OrderReturn/OrderReturn.tsx

```diff
@@ -87,11 +76,86 @@ const OrderReturn: React.FC<OrderReturnProps> = ({ orderId }) => {
         },
       }),
     );
+
+    if (returnErrors.length) {
+      return;
+    }
+
+    const grantRefundData = formData.autoGrantRefund
+      ? await grantRefund({
+          variables: {
+            orderId,
+            amount: formData.amount,
+            reason: "",
+            lines: [
+              ...formData.fulfilledItemsQuantities.map(line => ({
+                id: line.data.orderLineId,
+                quantity: line.value,
+              })),
+              ...formData.unfulfilledItemsQuantities.map(({ id, value }) => ({
+                id,
+                quantity: value,
+              })),
+            ],
+            grantRefundForShipping: formData.refundShipmentCosts,
+          },
+        })
+      : null;
+
+    const grantRefundErrors =
+      grantRefundData?.data?.orderGrantRefundCreate?.errors ?? [];
+
+    const grantedRefundId =
+      grantRefundData?.data?.orderGrantRefundCreate?.grantedRefund?.id;
+    const isSendRefund = grantedRefundId && formData.autoSendRefund;
+    const sendRefundErrors = isSendRefund
+      ? await extractMutationErrors(
+          sendRefund({
+            variables: {
+              grantedRefundId,
+              transactionId: data?.order?.transactions[0].id,
+            },
+          }),
+        )
+      : [];
```

Comment:
It's too complex:

multiple inline spreads that are hard to read
long inline call within the component
It can be hook that manages grand refund and returns the state along with the data taking form as an input